### ISSUE TITLE
Support non-node environments

### ISSUE THREAD

**Issue by pushrax** [turbopuffer/turbopuffer-typescript#20]
Currently the HTTP backend has a direct dependency on node.js so we can fine tune performance.

However, this makes it incompatible with non-node environments.

The HTTP backend is pluggable (see [httpClient.ts](https://github.com/turbopuffer/turbopuffer-typescript/blob/94595ae053cfb4b00944f3b62389d349408a4ae2/src/httpClient.ts)). We should be able to make another backend implementation that uses [isomorphic-fetch](https://github.com/matthew-andrews/isomorphic-fetch) to support all other environments that have `fetch`.

Essentially this boils down to:
- move DefaultHTTPClient from httpClient.ts into another file (e.g. nodeHttpClient.ts), as well as any related `undici` imports
- copy DefaultHTTPClient to our new client file (e.g. fetchHttpClient.ts), import `isomorphic-fetch`, and fix any type errors that pop up (should be a relatively easy switch)
- (depending on how the packaging works)
    a. change createHTTPClient to import the correct client for the environment and return it, or
    b. make a couple stub files that are conditionally imported by packaging flags, and are used only to plumb up createHTTPClient based on the environment

I'm not very familiar with how JS packaging works, and whether we should handle multiple environments by using multiple packages, or just using one package and doing some kind of conditional imports.

**Comment by pushrax** [comment]
https://github.com/openai/openai-node/tree/master/src/_shims is a good reference

**Comment by fforres** [comment]
Hi folks, was coming here to submit the same issue as I'm looking to implement turbopuffer in a non node-js environment (Specifically cloudflare workers). 
Having the same issues w/ `httpClient.js` and `undici`/`fetch` 

(JIC, among other things, cloudflare workers/workerd environments do not support WeakMap https://github.com/cloudflare/workerd/issues/3053#issuecomment-2456723399 which undici's fetch relies on)

**Comment by sirupsen** [comment]
@fforres this is pretty high on our list, but unlikely the tpuf team will be able to do a good patch for this in the next 1-2 weeks. 😭 

If you have cycles to submit a patch, we'd ❤ it

**Comment by fforres** [comment]
Will try to prepare something up @sirupsen 

Though have a question as this does feel more of a design question on your folk's end. 

While folks could use something like `node:http`/`node:https` for runtimes like cloudflare workers, While the modules are somewhat polyfilled the `unenv` underlying library does not allow for http/https.request, so that's a no-go sadly.

https://github.com/unjs/unenv/blob/main/src/runtime/node/https/index.ts#L17-L18
https://github.com/unjs/unenv/blob/main/src/runtime/node/http/index.ts#L15


On the other hand. `fetch` spec does not allow for keep-alive/agents. 


TLDR. For cloudlfare/edge-envs (ephemeral/stateless) using fetch makes sense, but unsure of what you folks are planning on here?

**Comment by sirupsen** [comment]
@fforres I think we should use whatever supports keep-alive, and failing that, yeah, fetch. I think Cloudflare's runtime at least will do keep-alive in the background, but unsure of others like Bun/Node

**Comment by pushrax** [comment]
As my original comment indicates, we need to have separate backends for different runtimes. We should continue using our existing undici backend for Node, it's heavily optimized.

All other runtimes should just use native fetch()

Bun tries to have API compatibility with Node but it's missing support for the undici APIs we use, so it'll need to take the less optimized path for now.
